ユーザのフォロー、アンフォロー機能の実装

goruchanchan opened the pull request:
ご確認よろしくお願いします🙇‍♂️

goruchanchan committed:
add 画像のアップロードと表示機能の実装

とりあえず上げた画像を表示できるところまで実装。リサイズなどはしていないので今後対応

goruchanchan committed:
update 画像のリサイズ対応

goruchanchan committed:
update 編集画面にアバターを追加

goruchanchan committed:
update i18n対応

goruchanchan committed:
clean rubocop対応

goruchanchan committed:
add erblint対応

goruchanchan committed:
remove プラクティスで使用しなかったファイルの削除

goruchanchan committed:
add erb-lint対応

goruchanchan committed:
update .erb-lint.ymlの内容修正

Sinatraのものを記述していたので修正

goruchanchan committed:
add フォロー用テーブルの追加

goruchanchan committed:
add データモデルの関係を定義

goruchanchan committed:
clean アイテマイズ整理

goruchanchan committed:
merged followブランチで作業していたつもりが、画像追加ブランチで作業していたので、作業分をマージ

Merge branch 'my-user_icon' into my-follow_users

goruchanchan committed:
clean rubocop対応

goruchanchan committed:
update 不要ファイルの削除

validateはテスト用なので、今回は不要

goruchanchan committed:
add フォロー一覧、フォロワー一覧の追加

goruchanchan committed:
add フォロー機能の導入

アンフォローを入れたつもりだが機能していないので要修正
destroyへのパスが通っていない感じな気がする

goruchanchan committed:
add アンフォロー機能実装

goruchanchan committed:
clean rubocop対応

goruchanchan committed:
clean 不要コード削除

goruchanchan committed:
clean 不要コメントの削除

なくてもわかるメソッドのコメントを削除

goruchanchan committed:
update 変更不要だった部分を元のブランチの内容に戻した

betachelsea commented on db/seeds.rb:
```diff
@@ -52,27 +52,34 @@ def picture_file(name)
-end
+# 開発用ユーザー
+User.create!(
+  email: 'hoge@mail.com',
```
[MUST] `mail.com` というのは実際に存在するプロバイダーのドメインなので、いつか実際の送信周りの実装を書いたとき、うっかりどこかのタイミングで本当の送信作業に使った時に誤って知らない人に送ってしまう...という問題につながる可能性がありますね。 :eyes: 
`example.com` というドメインは、そういうものに使っても良いというルールになっているドメインなので、仮のサンプルメールアドレスを記載する場合は積極的に `example.com` を使用してください！ 👍 


[example\.comとは \- 意味をわかりやすく \- IT用語辞典 e\-Words](https://e-words.jp/w/example.com.html#:~:text=example.com%E3%81%A8%E3%81%AF%E3%80%81%E4%BE%8B%E7%A4%BA,%E3%82%88%E3%81%86%E4%BF%9D%E8%AD%B7%E3%81%95%E3%82%8C%E3%81%A6%E3%81%84%E3%82%8B%E3%80%82)

goruchanchan replied:
本当に送信されてしまう恐れ大ですね😱確かに、`seeds.rb`で登録される他のアドレスは全部 `example.com` とされてますね😲

参考リンクもありがとうございます！

> 例示用に予約されたドメイン名の一つ

腹落ちしました🙇‍♂️

betachelsea commented on config/routes.rb:
```diff
@@ -4,4 +4,10 @@
+      get :following, :followers
+    end
+  end
+  resources :follows, only: [:create, :destroy]
```
[nits] 現在のpathは次のようになっている様子ですね。

-  POST /follows
-  DELETE /follows/:id

こちら、Controller側で id のもらい方が create/destroyで異なり、やりにくかった...という感じではありませんでしたか？ 
うまく「ネストを活用」・「単数系リソースを使用」することで、どちらのpathの形もうまく揃えることができそうです。
もし余力あれば、試してみていただきたいです〜 🙏  (動作に問題は無いので、必須修正ではないかなと考えます)

[Rails のルーティング \- Railsガイド \- 2.5 単数形リソース](https://railsguides.jp/routing.html#%E5%8D%98%E6%95%B0%E5%BD%A2%E3%83%AA%E3%82%BD%E3%83%BC%E3%82%B9)

betachelsea commented on app/controllers/users_controller.rb:
```diff
@@ -1,11 +1,30 @@
+  def show; end
+
+  def following
+    @title = User.human_attribute_name(:follow)
```
[nits] viewに表記する細かい差異について、都度インスタンス変数を定義していくと、開発が進むにつれて管理が大変になってきます。すこしずつ違う表記にしたい場合、パーシャル view を使って共通化するのをオススメします。ぜひやってみてください！

今回のケースでいうと、
- `app/views/users/following.html.erb` と `app/views/users/followers.html.erb` をそれぞれ作る
-  パーシャル（部分テンプレート）を作り、それぞれのviewから呼び出す

という形にすると、管理するインスタンス変数の数を減らせそうだなと思いました 👀 

[レイアウトとレンダリング \- Railsガイド - 3.4 パーシャルを使う](https://railsguides.jp/layouts_and_rendering.html#%E3%83%91%E3%83%BC%E3%82%B7%E3%83%A3%E3%83%AB%E3%82%92%E4%BD%BF%E3%81%86)

betachelsea commented on app/models/user.rb:
```diff
@@ -3,6 +3,30 @@
+                             inverse_of: :user,
+                             dependent: :destroy
+
+  has_many :following, through: :active_follows, source: :followed_user
```
[nits] `followers` に対して、 `following` が単数系なのが気になりました〜 :eyes:  (他の箇所についての命名についても同様のことが言える箇所がありそうです)

betachelsea commented on app/views/users/show.html.erb:
```diff
@@ -29,10 +29,28 @@
+
+<p>
+  <strong><%= User.human_attribute_name(:follower) %>:</strong>
+  <a href="<%= followers_user_path(@user) %>">
```
[nits] よりRailsっぽさを求めると、 `<a href="..."></a>` は `<%= link_to ... %>` を使用すると良いですね！

betachelsea commented on app/models/user.rb:
```diff
@@ -3,6 +3,30 @@
+  end
+
+  def unfollow(other_user)
+    active_follows.find_by(followed_user_id: other_user.id).destroy
```
[MUST] follow, unfollow の操作それぞれで、次のようにすると例外エラーが発生します。

1. 特定のユーザのページを複数タブ(A,Bとする)で開く
2. Aのタブで「フォローする」ボタンを押す。フォローが正常にできる。
3. Bのタブで「フォローする」ボタンを押す。例外エラーが発生する。

※「フォロー解除」についても、エラーは少し違うかもしれないが、同様の問題が発生する。

こちら対処してみてもらいたいです！ 🙏

betachelsea reviewed (commented):
コメントしました！ご確認よろしくお願いします〜 📝

goruchanchan committed:
サンプルメールアドレスのドメイン変更

サンプル用のメールアドレスを利用する際は、誤送を避けるために実在しないことが保証されているexamle.comドメインを使用する

goruchanchan committed:
update followsコントローラへのパスの統一

goruchanchan committed:
update パーシャルによるインスタンス変数削減

goruchanchan committed:
update 一対多関係の属性名を複数形に変更

goruchanchan committed:
update aタグをlink_toメソッドに変更

goruchanchan committed:
update 不要なレンダーを削除

誤って紛れ込んでしまっていたレンダーを削除

goruchanchan committed:
fix 複数タブで特定ユーザを、それぞれでフォロー・アンフォローした際に発生するバグ修正

goruchanchan committed:
update モデルのメソッド修正

他で使用することもないので、わざわざモデル内に定義不要と考え、ビューに移動

goruchanchan commented:
@betachelsea 

ご確認ありがとうございます！指摘事項を修正しました！再度レビューお願いします！

betachelsea commented on config/routes.rb:
```diff
@@ -4,4 +4,15 @@
+      get :followings, :followers
+    end
+  end
+  resources :follows, only: %i() do
+    member do
+      post :create
+      delete :destroy
+    end
+  end
```
良い感じに `/follows/:id` で POST, DELETE を受け付けるようになりましたね！ 👍 

[IMO] ユーザーの 

- `GET /users/:id/followings` 
- `GET /users/:id/followers`

があるので、その流れで次のようにするのも一手かなと思ったので、共有します〜 🙏 

- `POST /users/:user_id/follows`
- `DELETE /users/:user_id/follows`

こうすることのメリットとして、ルーティングが次のようにかけるので、スッキリできるという点があげられます。

```ruby
resources :users do
  resource :follows, only: %i(create destroy)
  member do
    get :followings, :followers
  end
end
```

betachelsea commented on config/routes.rb:
```diff
@@ -4,4 +4,15 @@
   root to: 'books#index'
   resources :books
   resources :users, only: %i(index show)
+  resources :users do
```
[nits] ここはひとまとまりにするのが適切そうですね！ 上の行でせっかく `only` でしぼっていますが、下の行で改めて7つのactionが定義されてしまっているようです :eyes: 

```ruby
resources :users, only: %i(index show) do
   # ...略...
end
```

betachelsea commented on app/views/users/_index_follow.html.erb:
```diff
@@ -0,0 +1,31 @@
+  </thead>
+
+  <tbody>
+    <% @users.each do |user| %>
```
良い感じのパーシャルビューですね！ 👍 

[MUST] 1点、せっかくパーシャルビューを活用いただいたので、調整していただきたいところがあります。
ここで `@users` のインスタンス変数の呼び出しがありますが、こうすると、「このパーシャルビューを使う場合、必ず事前に `@users` を定義しないといけない」という縛り(依存)が生まれてしまいます。
これの回避方法として、呼び出す場合に

```ruby
<%= render "index_follow", users: @users %>
```

とすると、

```ruby
# app/views/users/_index_follow.html.erb

<% users.each do |user| %>
  ...中略...
<% end %>
```

と、普通の変数として扱うことができるようになります。こちらの形になるよう、見直ししてみてもらえますでしょうか！ 🙏

goruchanchan replied:
> 事前に @users を定義しないといけない

上記のように変更してあげれば、インスタンス変数ではなく、呼び出し時に用意した変数を引数として渡せる。
よって、パーシャルビューとしては、『インスタンス変数である必要はなくなり、何かしらの変数が呼び出し側で用意されていれば良くなる』という理解で良いでしょうか🧐

betachelsea replied:
@goruchanchan はい、その理解で大丈夫だと思いました！ 🙆

betachelsea commented on app/models/user.rb:
```diff
@@ -3,6 +3,26 @@
+  end
+
+  def unfollow(other_user)
+    active_follows.find_by(followed_user_id: other_user.id).destroy if followings.include?(other_user)
```
問題に対処できていますね！ ありがとうございます！👍 

[FYI] 別解として、このdestroy時の問題については、「`find_by` した結果が無い場合、nilを返し、nilに対してdestroyは呼べないので例外」という要因でした。ここで、 nilに対してとりあえず呼び出しをかけることができる方法として、

```ruby
active_follows.find_by(followed_user_id: other_user.id)&.destroy 
```

という書き方もあるので、共有しておきます〜 🙏 
この `&.` は safe navigation演算子(または、"Null条件演算子"、Rubyの書き方により"ぼっち演算子"とも) というもので、「nilに対して呼び出しがあったときはnilを返す」という動きをしてくれて、便利です 😄 

参考: [Rubyのぼっち演算子はRailsの\`Object\#try\`より高速（翻訳）｜TechRacho by BPS株式会社](https://techracho.bpsinc.jp/hachi8833/2018_03_30/52987)

goruchanchan replied:
チェリー本にもでてきた『ぼっち演算子』😳

出会うのはもっと先かと思っていましたが、こういうところで使うんですね😲
『Null条件演算子』という呼び方は、名が体を表していて覚えやすいと思いました🧐

betachelsea commented on app/controllers/users_controller.rb:
```diff
@@ -1,11 +1,28 @@
+
+  def followings
+    @users = @user.followings
+    render 'followings'
```
[nits] renderは暗黙的にアクション名と同じviewを呼びにいってくれるので、この行は無くても動きそうです〜 👀

betachelsea reviewed (commented):
修正ありがとうございます！ せっかくなのでもうちょっと見てほしいなという部分があり、コメントさせていただきました〜 🙏

goruchanchan committed:
update usersモデルのルート設定変更

onlyで絞ったのに、actionを追加してしまっていたので修正

goruchanchan committed:
update routes整理

followモデルをuserモデルのroutesに含めることで、ルーティングを簡潔書けるようにした

goruchanchan committed:
update 不要なrenderの削除

goruchanchan committed:
update パーシャルビューでインスタンス変数ではなく、引数として渡すように変更

上記のようにすることで、インスタンス変数という縛りではなく、呼び出しがわで何かしら用意すればパーシャルビューが使えるようになるので依存性が低くなる。

goruchanchan committed:
update ぼっち演算子を使う方法に変更

goruchanchan committed:
clean rubocop対応

betachelsea commented on app/models/user.rb:
```diff
@@ -3,6 +3,25 @@
+  has_many :followers, through: :passive_follows, source: :user
+
+  def follow(other_user)
+    followings << other_user unless followings.include?(other_user)
```
[FYI] `unfollow` のほうはぼっち演算子を活用していただき、ありがとうございます！ 👍 
こちらの `followings << other_user` の記法について、ユニークのエラーを回避するには、書かれているように `unless` で既存のレコード有無を判断するしかなさそうですね。
補足として、こちらもぼっち演算子の活用と同じ方向性の「実行しても問題がないようにする」形として、`find_or_create_by` メソッドを使うという方法もありますので、ご紹介しておきます 🙏 

```ruby
active_follows.find_or_create_by(followed_user: other_user)
```

betachelsea commented on app/controllers/users_controller.rb:
```diff
@@ -1,11 +1,26 @@
+  def show; end
+
+  def followings
+    @users = @user.followings
```
[nits]  `@users` のインスタンス変数に取る際、`@user.followings` のみだと partial viewの変数に与えるときにそのまま `@user.followings` を書いて良いのではという気持ちになるので、セットで下記のような工夫をしておくと、`@users` 変数を活用する意義が強まりそうだなと思いました！ :eyes: 

- ページングの対応
- 表示順が確実になるような対応
- N+1への対応(画像の表示に対して) 例: [Active StorageのN\+1問題に対処する \- シュッと開発日記](https://shuttodev.hatenablog.com/entry/2019/09/10/012916)

betachelsea reviewed (commented):
修正ありがとうございます！確認OKです 🙆 
少し補足コメントを追加しました〜 📝

goruchanchan committed:
update find_or_create_by メソッドの導入

unlessは実行しないという対処だったが、find_or_create_byでは実行されても問題ないという対処方法。
方向性として、find_* で書いておいた方が統一感があるのでこっちを採用。

goruchanchan committed:
update ページング対応
